libcudacxx/include/cuda/std/detail/libcxx/include/__cuda/memcpy_async.h

```diff
@@ -0,0 +1,525 @@
+    return { _CUDA_VSTD::forward<_Group>(__g), __out_ptr, __in_ptr, __size, __sync };
+}
+
+template<bool _Tx, _CUDA_VSTD::size_t _NativeAlignment, typename _Arch, typename _Group, typename _Size, typename _SyncObject>
+struct __memcpy_async_space_dispatcher_t {
+    _Group && __g;
+    char * __out_ptr;
+    const char * __in_ptr;
+    _Size __size;
+    _SyncObject & __sync;
+
+    template<typename _OutSpace, typename _InSpace, typename _SyncSpace>
+    _LIBCUDACXX_INLINE_VISIBILITY
+    async_contract_fulfillment operator()(_OutSpace, _InSpace, _SyncSpace) {
+        using __hooks = __memcpy_async_hooks<_Tx,
+            _Arch,
+            _OutSpace::value,
+            _InSpace::value,
+            _SyncSpace::value>;
+
+        auto __f = __memcpy_async_alignment_dispatcher<__hooks, _NativeAlignment, _Arch>(_CUDA_VSTD::forward<_Group>(__g), __out_ptr, __in_ptr, __size, __sync);
+
+        auto __alignment_bit_v = __hooks::__compute_alignment_bit(__g, __out_ptr, __in_ptr, __size, __sync);
+
+        auto __acf = __dispatch_alignment_bit<
+            __hooks::__max_interesting_alignment,
+            __hooks::__min_interesting_alignment,
+            (_NativeAlignment > decltype(__get_alignment(__size))::value ? _NativeAlignment : decltype(__get_alignment(__size))::value)
+        >(__f, __alignment_bit_v);
+
+        using __sync_hooks = __memcpy_async_sync_hooks<_CUDA_VSTD::__remove_cvref_t<_SyncObject>,
+            _Tx,
+            _Arch,
+            _OutSpace::value,
+            _InSpace::value,
+            _SyncSpace::value>;
+        return __sync_hooks::__synchronize(_Arch{}, __sync, __acf);
+    }
+};
+
+template<bool _Tx, _CUDA_VSTD::size_t _NativeAlignment, typename _Arch, typename _Group, typename _Size, typename _SyncObject>
+_LIBCUDACXX_INLINE_VISIBILITY
+__memcpy_async_space_dispatcher_t<_Tx, _NativeAlignment, _Arch, _Group, _Size, _SyncObject> __memcpy_async_space_dispatcher(_Group && __g, char * __out_ptr, const char * __in_ptr, _Size __size, _SyncObject & __sync) {
+    return { _CUDA_VSTD::forward<_Group>(__g), __out_ptr, __in_ptr, __size, __sync };
+}
+
+template<bool _Tx, _CUDA_VSTD::size_t _NativeAlignment, typename _Group, typename _Size, typename _SyncObject>
+struct __memcpy_async_arch_dispatcher_t {
+    _Group && __g;
+    char * __out_ptr;
+    const char * __in_ptr;
+    _Size __size;
+    _SyncObject & __sync;
+
+    template<typename _Arch>
+    _LIBCUDACXX_INLINE_VISIBILITY
+    async_contract_fulfillment operator()(_Arch) {
+        auto __f = __memcpy_async_space_dispatcher<_Tx, _NativeAlignment, _Arch>(_CUDA_VSTD::forward<_Group>(__g), __out_ptr, __in_ptr, __size, __sync);
+
+        auto __sync_ptr = &__sync;
+        _LIBCUDACXX_HANDLE_POINTER_SPACE(__out_ptr, _LIBCUDACXX_HANDLE_POINTER_SPACE(__in_ptr, _LIBCUDACXX_HANDLE_POINTER_SPACE(__sync_ptr,
```

Comment:
🙀

Reply:
Trust me, this is easier to work with than a wrapper that I started with.